File: sdk/python/src/protomcp/middleware.py
from dataclasses import dataclass
from typing import Callable, Any

_middleware_registry: list["MiddlewareDef"] = []

@dataclass
class MiddlewareDef:
    name: str
    priority: int
    handler: Callable  # (phase, tool_name, args_json, result_json, is_error) -> dict

def middleware(name: str, priority: int = 100):
    def decorator(func: Callable) -> Callable:
        _middleware_registry.append(MiddlewareDef(
            name=name,
            priority=priority,
            handler=func,
        ))
        return func
    return decorator

def get_registered_middleware() -> list[MiddlewareDef]:
    return list(_middleware_registry)

def clear_middleware_registry():
    _middleware_registry.clear()

File: sdk/go/protomcp/middleware.go
package protomcp

import "sync"

// MiddlewareDef holds a registered middleware definition.
type MiddlewareDef struct {
	Name     string
	Priority int32
	Handler  func(phase, toolName, argsJSON, resultJSON string, isError bool) map[string]interface{}
}

var (
	middlewareRegistry []MiddlewareDef
	middlewareMu       sync.Mutex
)

// Middleware registers a custom middleware handler that intercepts tool calls.
// Priority determines execution order: lower numbers run first in the before phase.
func Middleware(name string, priority int32, handler func(phase, toolName, argsJSON, resultJSON string, isError bool) map[string]interface{}) {
	middlewareMu.Lock()
	defer middlewareMu.Unlock()
	middlewareRegistry = append(middlewareRegistry, MiddlewareDef{
		Name:     name,
		Priority: priority,
		Handler:  handler,
	})
}

// GetRegisteredMiddleware returns a copy of the middleware registry.
func GetRegisteredMiddleware() []MiddlewareDef {
	middlewareMu.Lock()
	defer middlewareMu.Unlock()
	result := make([]MiddlewareDef, len(middlewareRegistry))
	copy(result, middlewareRegistry)
	return result
}

// ClearMiddlewareRegistry removes all registered middleware.
func ClearMiddlewareRegistry() {
	middlewareMu.Lock()
	defer middlewareMu.Unlock()
	middlewareRegistry = nil
}
